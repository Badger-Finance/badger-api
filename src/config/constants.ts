import { ethers } from 'ethers';

const RAW_TOKENS = {
  // eth tokens
  BADGER: '0x3472a5a71965499acd81997a54bba8d852c6e53d',
  DIGG: '0x798d1be841a82a273720ce31c822c61a67a601c3',
  WBTC: '0x2260fac5e5542a773aa44fbcfedf7c193bc2c599',
  WETH: '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2',
  SUSHI: '0x6b3595068778dd592e39a122f4f5a5cf09c90fe2',

  // bsc tokens
  CAKE: '0x0E09FaBB73Bd3Ade0a17ECC321fD13a19e81cE82',
  WBNB: '0xbb4CdB9CBd36B01bD1cBaEBF2De08d9173bc095c',
  BTCB: '0x7130d2A12B9BCbFAe4f2634d864A1Ee1Ce3Ead9c',
  BSC_BADGER: '0x753fbc5800a8C8e3Fb6DC6415810d627A387Dfc9',
  BSC_BBADGER: '0x1F7216fdB338247512Ec99715587bb97BBf96eae',
  BSC_BDIGG: '0x5986D5c77c65e5801a5cAa4fAE80089f870A71dA',

  // curve tokens
  CRV_RENBTC: '0x49849c98ae39fff122806c06791fa73784fb3675',
  CRV_TBTC: '0x64eda51d3ad40d56b9dfc5554e06f94e1dd786fd',
  CRV_SBTC: '0x075b1bb99792c9e1041ba13afef80c91a1e70fb3',

  // uniswap tokens
  UNI_BADGER_WBTC: '0xcd7989894bc033581532d2cd88da5db0a4b12859',
  UNI_DIGG_WBTC: '0xe86204c4eddd2f70ee00ead6805f917671f56c52',

  // sushiswap tokens
  SUSHI_ETH_WBTC: '0xceff51756c56ceffca006cd410b03ffc46dd3a58',
  SUSHI_BADGER_WBTC: '0x110492b31c59716ac47337e616804e3e3adc0b4a',
  SUSHI_DIGG_WBTC: '0x9a13867048e01c663ce8ce2fe0cdae69ff9f35e3',

  // pancakeswap tokens
  PANCAKE_BNB_BTCB: '0x356b7d0d3c54F22C82B7a670C6Ba9E2381b0624c',
  PANCAKE_BBADGER_BTCB: '0x87Ae7b5c43D4e160cDB9427a78BA87B9503ee37b',
  PANCAKE_BDIGG_BTCB: '0xfbfa92e037e37F946c0105902640914E3aCe6752',

  // eth vault tokens
  BBADGER: '0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28',
  BDIGG: '0x7e7e112a68d8d2e221e11047a72ffc1065c38e1a',
  BUNI_BADGER_WBTC: '0x235c9e24d3fb2fafd58a2e49d454fdcd2dbf7ff1',
  BUNI_DIGG_WBTC: '0xc17078fdd324cc473f8175dc5290fae5f2e84714',
  BSUSHI_ETH_WBTC: '0x758a43ee2bff8230eeb784879cdcff4828f2544d',
  BSUSHI_BADGER_WBTC: '0x1862a18181346ebd9edaf800804f89190def24a5',
  BSUSHI_DIGG_WBTC: '0x88128580acdd9c04ce47afce196875747bf2a9f6',
  BCRV_SBTC: '0xd04c48a53c111300ad41190d63681ed3dad998ec',
  BCRV_RENBTC: '0x6def55d2e18486b9ddfaa075bc4e4ee0b28c1545',
  BCRV_TBTC: '0xb9d076fde463dbc9f915e5392f807315bf940334',
  BCRV_HRENBTC: '0xaf5a1decfa95baf63e0084a35c62592b774a2a87',
  BVYWBTC: '0x4b92d19c11435614cd49af1b589001b7c08cd4d5',

  // bsc vault tokens
  BPANCAKE_BNB_BTCB: '0xaf4B9C4b545D5324904bAa15e29796D2E2f90813',
  BPANCAKE_BBADGER_BTCB: '0x857F91f735f4B03b19D2b5c6E476C73DB8241F55',
  BPANCAKE_BDIGG_BTCB: '0xa861Ba302674b08f7F2F24381b705870521DDfed',
};

export const TOKENS = Object.fromEntries(
  Object.entries(RAW_TOKENS).map(([key, val]) => [key, ethers.utils.getAddress(val)]),
);

export enum Protocol {
  Curve = 'curve',
  Sushiswap = 'sushiswap',
  Uniswap = 'uniswap',
  Pancakeswap = 'pancakeswap',
  Yearn = 'yearn',
}

// data point constants - index two times per hour, 48 per day
export const CURRENT = 0;
export const ONE_DAY = 24 * 2;
export const ONE_MINUTE_MS = 60 * 1000;
export const THREE_DAYS = ONE_DAY * 3;
export const SEVEN_DAYS = ONE_DAY * 7;
export const THIRTY_DAYS = ONE_DAY * 30;
export const SAMPLE_DAYS = THIRTY_DAYS + 1;
export const ONE_YEAR_MS = 365 * 24 * 60 * 60 * 1000;

// data access constants
export const APY_SNAPSHOTS_DATA = process.env.APY_SNAPSHOTS_DATA || 'MISSING REQUIRED ENV VAR';
export const ASSET_DATA = process.env.ASSET_DATA || 'MISSING REQUIRED ENV VAR';
export const PRICE_DATA = process.env.PRICE_DATA || 'MISSING REQUIRED ENV VAR';
export const SETT_SNAPSHOTS_DATA = process.env.SETT_SNAPSHOTS_DATA || 'MISSING REQUIRED ENV VAR';
export const REWARD_DATA = process.env.REWARD_DATA || 'MISSING REQUIRED ENV VAR';
export const UNISWAP_URL = process.env.UNISWAP || 'MISSING REQUIRED ENV VAR';
export const SUSHISWAP_URL = process.env.SUSHISWAP || 'MISSING REQUIRED ENV VAR';
export const PANCAKESWAP_URL = process.env.PANCAKESWAP || 'MISSING REQUIRED ENV VAR';
export const BADGER_URL = process.env.BADGER || 'MISSING REQUIRED ENV VAR';
export const BADGER_BSC_URL = process.env.BADGER_BSC || 'MISSING REQUIRED ENV VAR';
export const BADGER_DAO_URL = process.env.BADGER_DAO || 'MISSING REQUIRED ENV VAR';
export const MASTERCHEF_URL = process.env.MASTERCHEF || 'MISSING REQUIRED ENV VAR';
export const BOUNCER_PROOFS = process.env.BOUNCER_PROOFS || 'MISSING REQUIRED ENV VAR';

// third party api constants
export const CURVE_API_URL = 'https://stats.curve.fi/raw-stats/apys.json';
export const COINGECKO_URL = 'https://api.coingecko.com/api/v3/simple';

export enum Provider {
  Cloudflare = 'https://cloudflare-eth.com/',
  Binance = 'https://bsc-dataseed.binance.org/',
  MyEtherWallet = 'https://nodes.mewapi.io/rpc/eth',
}
export const SUSHI_CHEF = '0xc2EdaD668740f1aA35E4D8f227fB8E17dcA888Cd';
export const PANCAKE_CHEF = '0x73feaa1eE314F8c655E354234017bE2193C9E24E';
