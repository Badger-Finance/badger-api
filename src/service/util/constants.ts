export const TOKENS = {
	// general tokens
	RENBTC: '0x49849c98ae39fff122806c06791fa73784fb3675',
	BADGER: '0x3472a5a71965499acd81997a54bba8d852c6e53d',
	TBTC: '0x64eda51d3ad40d56b9dfc5554e06f94e1dd786fd',
	SBTC: '0x075b1bb99792c9e1041ba13afef80c91a1e70fb3',
	DIGG: '0x798d1be841a82a273720ce31c822c61a67a601c3',

	// uniswap tokens
	UNI_BADGER: '0xcd7989894bc033581532d2cd88da5db0a4b12859',
	UNI_DIGG: '0xe86204c4eddd2f70ee00ead6805f917671f56c52',

	// sushiswap tokens
	SUSHI_WBTC: '0xceff51756c56ceffca006cd410b03ffc46dd3a58',
	SUSHI_BADGER: '0x110492b31c59716ac47337e616804e3e3adc0b4a',
	SUSHI_DIGG: '0x9a13867048e01c663ce8ce2fe0cdae69ff9f35e3',
};
