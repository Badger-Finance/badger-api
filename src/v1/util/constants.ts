import { ethers } from 'ethers';

export const TOKENS = {
  // general tokens
  RENBTC: '0x49849c98ae39fff122806c06791fa73784fb3675',
  BADGER: '0x3472a5a71965499acd81997a54bba8d852c6e53d',
  TBTC: '0x64eda51d3ad40d56b9dfc5554e06f94e1dd786fd',
  SBTC: '0x075b1bb99792c9e1041ba13afef80c91a1e70fb3',
  DIGG: '0x798d1be841a82a273720ce31c822c61a67a601c3',

  // uniswap tokens
  UNI_BADGER: '0xcd7989894bc033581532d2cd88da5db0a4b12859',
  UNI_DIGG: '0xe86204c4eddd2f70ee00ead6805f917671f56c52',

  // sushiswap tokens
  SUSHI_WBTC: '0xceff51756c56ceffca006cd410b03ffc46dd3a58',
  SUSHI_BADGER: '0x110492b31c59716ac47337e616804e3e3adc0b4a',
  SUSHI_DIGG: '0x9a13867048e01c663ce8ce2fe0cdae69ff9f35e3',
};

// data point constants - index two times per hour, 48 per day
export const CURRENT = 0;
export const ONE_DAY = 24 * 2;
export const THREE_DAYS = ONE_DAY * 3;
export const SEVEN_DAYS = ONE_DAY * 7;
export const THIRTY_DAYS = ONE_DAY * 30;
export const SAMPLE_DAYS = THIRTY_DAYS + 1;
export const ONE_YEAR_MS = 365 * 24 * 60 * 60 * 1000;

// data access constants
export const ASSET_DATA = process.env.ASSET_DATA || 'MISSING REQUIRED ENV VAR';
export const UNISWAP_URL = process.env.UNISWAP || 'MISSING REQUIRED ENV VAR';
export const SUSHISWAP_URL = process.env.SUSHISWAP || 'MISSING REQUIRED ENV VAR';
export const BADGER_URL = process.env.BADGER || 'MISSING REQUIRED ENV VAR';
export const BADGER_DAO_URL = process.env.BADGER_DAO_URL || 'MISSING REQUIRED ENV VAR';
export const MASTERCHEF_URL = process.env.MASTERCHEF || 'MISSING REQUIRED ENV VAR';

// third party api constants
export const CURVE_API_URL = 'https://stats.curve.fi/raw-stats/apys.json';
export const ETHERS_JSONRPC_PROVIDER = new ethers.providers.JsonRpcProvider('https://cloudflare-eth.com/');
