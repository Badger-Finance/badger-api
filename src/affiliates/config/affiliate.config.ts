import { BadRequest } from '@tsed/exceptions';
import { Chain } from '../../chains/config/chain.config';
import { Protocol } from '../../config/constants';
import { SettAffiliateData } from '../../setts/interfaces/sett-affiliate-data.interface';
import { SettDefinition } from '../../setts/interfaces/sett-definition.interface';

type Affiliates = { [protocol: string]: Affiliate };

export abstract class Affiliate {
  private static affiliates: Affiliates = {};

  readonly name: string;

  constructor(name: string) {
    this.name = name;
  }

  static register(protocol: Protocol, affiliate: Affiliate): void {
    this.affiliates[protocol] = affiliate;
  }

  static getAffiliate(affiliateData: SettAffiliateData): Affiliate {
    const affiliate = this.affiliates[affiliateData.protocol];
    if (!affiliate) {
      throw new BadRequest(`${affiliate} is not a affiliate`);
    }
    return affiliate;
  }

  abstract getAffiliateVaultData(chain: Chain, sett: SettDefinition): Promise<SettAffiliateData>;
}
