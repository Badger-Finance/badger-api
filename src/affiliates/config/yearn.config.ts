import { BigNumber, ethers } from 'ethers';
import { Chain } from '../../chains/config/chain.config';
import { guestListAbi } from '../../config/abi/guest-list.abi';
import { yearnAffiliateVaultWrapperAbi } from '../../config/abi/yearn-affiliate-vault-wrapper.abi';
import { Protocol } from '../../config/constants';
import { SettAffiliateData } from '../../setts/interfaces/sett-affiliate-data.interface';
import { SettDefinition } from '../../setts/interfaces/sett-definition.interface';
import { getToken } from '../../tokens/tokens.utils';
import { Affiliate } from './affiliate.config';

export class Yearn extends Affiliate {
  constructor() {
    super(Protocol.Yearn);
    Affiliate.register(Protocol.Yearn, this);
  }

  async getAffiliateVaultData(chain: Chain, sett: SettDefinition): Promise<SettAffiliateData> {
    const affiliateData: SettAffiliateData = {
      protocol: Protocol.Yearn,
    };
    if (sett.hasBouncer) {
      try {
        const vaultWrapper = new ethers.Contract(sett.settToken, yearnAffiliateVaultWrapperAbi, chain.provider);
        const guestListAddress = await vaultWrapper.guestList();
        const guesList = new ethers.Contract(guestListAddress, guestListAbi, chain.provider);
        const [available, limit]: [BigNumber, BigNumber] = await Promise.all([
          guesList.remainingTotalDepositAllowed(),
          guesList.totalDepositCap(),
        ]);
        const depositToken = getToken(sett.depositToken);
        affiliateData.availableDepositLimit = parseFloat(ethers.utils.formatUnits(available, depositToken.decimals));
        affiliateData.depositLimit = parseFloat(ethers.utils.formatUnits(limit, depositToken.decimals));
      } catch {
        affiliateData.availableDepositLimit = 0;
        affiliateData.depositLimit = 50;
      }
    }
    return affiliateData;
  }
}
