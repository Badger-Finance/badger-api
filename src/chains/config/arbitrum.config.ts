import { Protocol } from '../../config/enums/protocol.enum';
import { SettState } from '../../config/enums/sett-state.enum';
import { Stage } from '../../config/enums/stage.enum';
import rpc from '../../config/rpc.config';
import { TOKENS } from '../../config/tokens.config';
import { GasPrices } from '../../gas/interfaces/gas-prices.interface';
import { getCurveSettTokenBalance } from '../../protocols/strategies/convex.strategy';
import { BouncerType } from '../../rewards/enums/bouncer-type.enum';
import { SettDefinition } from '../../setts/interfaces/sett-definition.interface';
import { arbitrumTokensConfig } from '../../tokens/config/arbitrum-tokens.config';
import { ChainNetwork } from '../enums/chain-network.enum';
import { ArbitrumStrategy } from '../strategies/arbitrum.strategy';
import { Chain } from './chain.config';

export class Arbitrum extends Chain {
  constructor() {
    super(
      'Arbitrum',
      'arbitrum',
      '0xa4b1',
      ChainNetwork.Arbitrum,
      arbitrumTokensConfig,
      arbitrumSetts,
      rpc[ChainNetwork.Arbitrum],
      new ArbitrumStrategy(Object.keys(arbitrumTokensConfig)),
      2425847,
      '0x635EB2C39C75954bb53Ebc011BDC6AfAAcE115A6',
      '0x85E1cACAe9a63429394d68Db59E14af74143c61c',
    );
    Chain.register(this.network, this);
  }

  async getGasPrices(): Promise<GasPrices> {
    const gasPrice = 2;
    return {
      rapid: gasPrice,
      fast: gasPrice,
      standard: gasPrice,
      slow: gasPrice,
    };
  }
}

export const arbitrumSetts: SettDefinition[] = [
  {
    name: 'Sushiswap Wrapped Ether/Sushi Helper',
    settToken: TOKENS.BARB_SUSHI_WETH_SUSHI,
    depositToken: TOKENS.ARB_SUSHI_WETH_SUSHI,
    createdBlock: 13163959,
    protocol: Protocol.Sushiswap,
    strategy: '0x86f772C82914f5bFD168f99e208d0FC2C371e9C2',
  },
  {
    name: 'Sushiswap Wrapped BTC/Wrapped ETH',
    settToken: TOKENS.BARB_SUSHI_WETH_WBTC,
    depositToken: TOKENS.ARB_SUSHI_WETH_WBTC,
    createdBlock: 13163959,
    protocol: Protocol.Sushiswap,
    strategy: '0xA6827f0f14D0B83dB925B616d820434697328c22',
  },
  {
    name: 'Curve renBTC/wBTC',
    settToken: TOKENS.BARB_CRV_RENBTC,
    depositToken: TOKENS.ARB_CRV_RENBTC,
    getTokenBalance: getCurveSettTokenBalance,
    createdBlock: 13237551,
    protocol: Protocol.Curve,
  },
  {
    name: 'Curve Tricrypto',
    settToken: TOKENS.BARB_CRV_TRICRYPTO,
    depositToken: TOKENS.ARB_CRV_TRICRYPTO,
    getTokenBalance: getCurveSettTokenBalance,
    createdBlock: 13237551,
    protocol: Protocol.Curve,
  },
  {
    name: 'Curve Tricrypto Light',
    settToken: TOKENS.BARB_CRV_TRICRYPTO_LITE,
    depositToken: TOKENS.ARB_CRV_TRICRYPTO,
    getTokenBalance: getCurveSettTokenBalance,
    createdBlock: 13321375,
    stage: Stage.Staging,
    state: SettState.Experimental,
    protocol: Protocol.Curve,
  },
  {
    name: 'Swapr Swapr/Wrapped ETH',
    settToken: TOKENS.BARB_SWP_SWPR_WETH,
    depositToken: TOKENS.ARB_SWP_SWPR_WETH,
    createdBlock: 13315350,
    state: SettState.Guarded,
    protocol: Protocol.Swapr,
  },
  {
    name: 'Swapr Wrapped BTC/Wrapped ETH',
    settToken: TOKENS.BARB_SWP_WBTC_WETH,
    depositToken: TOKENS.ARB_SWP_WBTC_WETH,
    createdBlock: 13315350,
    state: SettState.Guarded,
    protocol: Protocol.Swapr,
  },
  {
    name: 'Swapr Badger/Wrapped ETH',
    bouncer: BouncerType.Badger,
    settToken: TOKENS.BARB_SWP_BADGER_WETH,
    depositToken: TOKENS.ARB_SWP_BADGER_WETH,
    createdBlock: 2188169,
    state: SettState.Guarded,
    protocol: Protocol.Swapr,
  },
];
