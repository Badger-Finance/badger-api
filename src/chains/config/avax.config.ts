import { Network, Protocol, VaultState, VaultVersion } from '@badger-dao/sdk';
import rpc from '../../config/rpc.config';
import { TOKENS } from '../../config/tokens.config';
import { GasPrices } from '../../gas/interfaces/gas-prices.interface';
import { VaultDefinition } from '../../vaults/interfaces/vault-definition.interface';
import { Chain } from './chain.config';
import { BaseStrategy } from '../strategies/base.strategy';
import { avalancheTokensConfig } from '../../tokens/config/avax-tokens.config';
import { Stage } from '../../config/enums/stage.enum';

export class Avalanche extends Chain {
  constructor() {
    super(
      'Avalanche',
      'avalanche',
      '0xa86a',
      Network.Avalanche,
      avalancheTokensConfig,
      rpc[Network.Avalanche],
      avalancheSetts,
      new BaseStrategy(Network.Avalanche, Object.keys(avalancheTokensConfig)),
    );
    Chain.register(this.network, this);
  }

  async getGasPrices(): Promise<GasPrices> {
    return this.defaultGasPrice();
  }

  // TODO: Update if badger is ever updated
  getBadgerTokenAddress(): string {
    return TOKENS.BADGER;
  }
}

export const avalancheSetts: VaultDefinition[] = [
  {
    name: 'WBTC',
    vaultToken: TOKENS.BAVAX_WBTC,
    depositToken: TOKENS.AVAX_WBTC,
    protocol: Protocol.Aave,
    stage: Stage.Staging,
    state: VaultState.Experimental,
    version: VaultVersion.v1_5,
  },
];
