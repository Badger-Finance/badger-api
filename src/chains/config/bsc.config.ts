import { Network } from "@badger-dao/sdk";

import rpc from "../../config/rpc.config";
import { TOKENS } from "../../config/tokens.config";
import { bscTokensConfig } from "../../tokens/config/bsc-tokens.config";
import { BaseStrategy } from "../strategies/base.strategy";
import { Chain } from "./chain.config";

export class BinanceSmartChain extends Chain {
  constructor() {
    super(Network.BinanceSmartChain, bscTokensConfig, rpc[Network.BinanceSmartChain], new BaseStrategy(Network.BinanceSmartChain));
    Chain.register(this.network, this);
  }

  getBadgerTokenAddress(): string {
    return TOKENS.MULTI_BADGER;
  }
}
