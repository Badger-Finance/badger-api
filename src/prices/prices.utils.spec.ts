import { DataMapper, QueryIterator, StringToAnyObjectMap } from '@aws/dynamodb-data-mapper';
import { BadRequest, NotFound } from '@tsed/exceptions';
import createMockInstance from 'jest-create-mock-instance';
import fetchMock from 'jest-fetch-mock';
import { TestStrategy } from '../chains/strategies/test.strategy';
import { TOKENS } from '../config/constants';
import { TokenType } from '../tokens/enums/token-type.enum';
import { Token } from '../tokens/interfaces/token.interface';
import { TokenPrice } from '../tokens/interfaces/token-price.interface';
import { TokenPriceSnapshot } from '../tokens/interfaces/token-price-snapshot.interface';
import { getToken, protocolTokens } from '../tokens/tokens.utils';
import {
  getContractPrice,
  getPrice,
  getPriceData,
  getTokenPrice,
  getTokenPriceData,
  getVaultTokenPrice,
  inCurrency,
  noPrice,
  updatePrice,
  updatePrices,
} from './prices.utils';

describe('prices.utils', () => {
  const strategy = new TestStrategy();

  // Father forgive me for I have sinned
  /* eslint-disable @typescript-eslint/ban-ts-comment */
  const setupMapper = (items: unknown[]) => {
    // @ts-ignore
    const qi: QueryIterator<StringToAnyObjectMap> = createMockInstance(QueryIterator);
    // @ts-ignore
    qi[Symbol.iterator] = jest.fn(() => items.values());
    return jest.spyOn(DataMapper.prototype, 'query').mockImplementation(() => qi);
  };
  /* eslint-enable @typescript-eslint/ban-ts-comment */

  beforeEach(async () => {
    fetchMock.resetMocks();
  });

  describe('getPrice', () => {
    describe('when price is not available', () => {
      it('returns a price of 0', async () => {
        const cake = getToken(TOKENS.CAKE);
        setupMapper([]);
        const cakePrice = await getPrice(cake.address);
        expect(cakePrice).toBeDefined();
        const expected = noPrice(cake);
        expected.updatedAt = cakePrice.updatedAt;
        expect(cakePrice).toMatchObject(expected);
      });
    });

    describe('when price is available', () => {
      it('returns a token snapshot with the latest price data', async () => {
        const price = {
          ...(await strategy.getPrice(TOKENS.BADGER)),
          updateAt: Date.now(),
        };
        setupMapper([price]);
        const fetchedBadgerPrice = await getPrice(TOKENS.BADGER);
        expect(fetchedBadgerPrice).toBeDefined();
        expect(fetchedBadgerPrice).toMatchObject(price);
      });
    });
  });

  describe('updatePrice', () => {
    describe('update unsupported token', () => {
      it('throws an bad request error', async () => {
        const unsupportedToken: Token = {
          name: 'Fake Token',
          symbol: 'FTK',
          address: '0xfA5047c9c78B8877af97BDcb85Db743fD7313d4a',
          decimals: 18,
          type: TokenType.Contract,
        };
        await expect(updatePrice(unsupportedToken)).rejects.toThrow(NotFound);
      });
    });

    describe('update supported token', () => {
      it('creates an price db entry', async () => {
        const badger = getToken(TOKENS.BADGER);
        const put = jest.spyOn(DataMapper.prototype, 'put').mockImplementation();
        await updatePrice(badger);
        expect(put.mock.calls.length).toEqual(1);
      });
    });
  });

  describe('updatePrices', () => {
    describe('update unsupported token', () => {
      it('throws an bad request error', async () => {
        const unsupportedToken: Token = {
          name: 'Fake Token',
          symbol: 'FTK',
          address: '0xfA5047c9c78B8877af97BDcb85Db743fD7313d4a',
          decimals: 18,
          type: TokenType.Contract,
        };
        await expect(updatePrices({ [unsupportedToken.address]: unsupportedToken })).rejects.toThrow(NotFound);
      });
    });

    describe('update a mix of unsupported and supported tokens', () => {
      it('throws a bad request error', async () => {
        const unsupportedToken: Token = {
          name: 'Fake Token',
          symbol: 'FTK',
          address: '0xfA5047c9c78B8877af97BDcb85Db743fD7313d4a',
          decimals: 18,
          type: TokenType.Contract,
        };
        const badgerToken = getToken(TOKENS.BADGER);
        const tokenConfig = {
          [unsupportedToken.address]: unsupportedToken,
          [TOKENS.BADGER]: badgerToken,
        };
        await expect(updatePrices(tokenConfig)).rejects.toThrow(NotFound);
      });
    });

    describe('update supported tokens', () => {
      it('updates prices for all tokens requested', async () => {
        const digg = getToken(TOKENS.DIGG);
        const badger = getToken(TOKENS.BADGER);
        const tokenConfig = {
          [digg.address]: digg,
          [badger.address]: badger,
        };
        const put = jest.spyOn(DataMapper.prototype, 'put').mockImplementation();
        await updatePrices(tokenConfig);
        expect(put.mock.calls.length).toEqual(Object.keys(tokenConfig).length);
        for (const call of put.mock.calls[0] as Iterable<TokenPriceSnapshot>) {
          expect(call.address).toBeDefined();
          const price = await strategy.getPrice(call.address);
          expect(call).toMatchObject(price);
        }
      });
    });
  });

  describe('getTokenPriceData', () => {
    describe('get unsupported token price', () => {
      it('throws a not found error', async () => {
        const unsupportedToken: Token = {
          name: 'Fake Token',
          symbol: 'FTK',
          address: '0xfA5047c9c78B8877af97BDcb85Db743fD7313d4a',
          decimals: 18,
          type: TokenType.Contract,
        };
        await expect(getTokenPriceData(unsupportedToken.address)).rejects.toThrow(NotFound);
      });
    });

    describe('get supported token price', () => {
      it('gets the latest price snapshot the from the db', async () => {
        const badger = getToken(TOKENS.BADGER);
        const mockPrice = {
          ...(await strategy.getPrice(badger.address)),
          updatedAt: Date.now(),
        };
        setupMapper([mockPrice]);
        const price = await getTokenPriceData(badger.address);
        expect(price).toMatchObject(mockPrice);
      });
    });
  });

  describe('getPriceData', () => {
    it('gets all token pricing for the system', async () => {
      const query = setupMapper([]);
      const data = await getPriceData(protocolTokens);
      expect(query.mock.calls.length).toEqual(Object.keys(data).length);
      const requests = [];
      /* eslint-disable @typescript-eslint/no-explicit-any */
      for (const call of query.mock.calls as Iterable<any>) {
        expect(call[1]).toBeDefined();
        const { address } = call[1];
        expect(address).toBeDefined();
        requests.push(address);
      }
      /* eslint-enable @typescript-eslint/no-explicit-any */
      expect(requests).toMatchObject(Object.keys(protocolTokens));
    });
  });

  describe('getContractPrice', () => {
    it('Fetches the contract price in USD and ETH', async () => {
      const contract = '0x3472A5A71965499acd81997a54BBA8D852C6E53d';
      const usdPrice = Math.random() * 100;
      const etherPrice = usdPrice / 1500;
      const mockResponse = {
        '0x3472a5a71965499acd81997a54bba8d852c6e53d': {
          usd: usdPrice,
          eth: etherPrice,
        },
      };
      fetchMock.mockResponseOnce(JSON.stringify(mockResponse));

      const response = await getContractPrice(contract);

      expect(fetchMock).toHaveBeenCalled();
      expect(response).toBeDefined();
      expect(response).toMatchObject({
        address: contract,
        usd: usdPrice,
        eth: etherPrice,
      });
    });
  });

  describe('getTokenPrice', () => {
    it('Fetches the token price in USD and ETH', async () => {
      const token = 'Badger';
      const usdPrice = Math.random() * 100;
      const etherPrice = usdPrice / 1500;
      const mockResponse = {
        Badger: {
          usd: usdPrice,
          eth: etherPrice,
        },
      };
      fetchMock.mockResponseOnce(JSON.stringify(mockResponse));

      const response = await getTokenPrice(token);

      expect(fetchMock).toHaveBeenCalled();
      expect(response).toBeDefined();
      expect(response).toMatchObject({
        name: token,
        usd: usdPrice,
        eth: etherPrice,
      });
    });
  });

  describe('inCurrency', () => {
    const priceData: TokenPrice = {
      name: 'Test',
      address: '0x0',
      usd: 10,
      eth: 10 / 1500,
    };

    describe('without a currency', () => {
      it('returns the usd price', () => {
        const defaultPrice = inCurrency(priceData);
        expect(defaultPrice).toEqual(priceData.usd);
      });
    });

    describe('with usd selected', () => {
      it('returns the usd price', () => {
        const defaultPrice = inCurrency(priceData, 'usd');
        expect(defaultPrice).toEqual(priceData.usd);
      });
    });

    describe('with eth selected', () => {
      it('returns the eth price', () => {
        const ethPrice = inCurrency(priceData, 'eth');
        expect(ethPrice).toEqual(priceData.eth);
      });
    });

    describe('with an unsupported currency', () => {
      it('returns the usd price', () => {
        const ethPrice = inCurrency(priceData, 'bchabc');
        expect(ethPrice).toEqual(priceData.usd);
      });
    });
  });

  describe('getVaultTokenPrice', () => {
    describe('look up non vault token price', () => {
      it('throws a bad request error', async () => {
        await expect(getVaultTokenPrice(TOKENS.BADGER)).rejects.toThrow(BadRequest);
      });
    });
  });
});
