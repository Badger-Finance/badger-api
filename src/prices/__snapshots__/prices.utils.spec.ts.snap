// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prices.utils fetchPrices request prices for contracts returns a price map of requested token prices in usd 1`] = `
Object {
  "0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599": Object {
    "usd": 43500,
  },
  "0x3472A5A71965499acd81997a54BBA8D852C6E53d": Object {
    "usd": 10,
  },
}
`;

exports[`prices.utils fetchPrices request prices for look up names returns a price map of requested token prices in usd 1`] = `
Object {
  "badger": Object {
    "usd": 10,
  },
  "wrapped-bitcoin": Object {
    "usd": 43500,
  },
}
`;
